csharp/src/Google.Protobuf/CodedInputReader.cs

```diff
@@ -0,0 +1,879 @@
+                ? reader.UnreadSpan.Slice(0, length).ToArray()
+                : reader.Sequence.Slice(reader.Position, length).ToArray();
+
+            reader.Advance(length);
+
+            return ByteString.AttachBytes(data);
+        }
+
+        /// <summary>
+        /// Reads a uint32 field value from the input.
+        /// </summary>   
+        public uint ReadUInt32()
+        {
+            return ReadRawVarint32();
+        }
+
+        /// <summary>
+        /// Reads an enum field value from the input.
+        /// </summary>   
+        public int ReadEnum()
+        {
+            // Currently just a pass-through, but it's nice to separate it logically from WriteInt32.
+            return (int)ReadRawVarint32();
+        }
+
+        /// <summary>
+        /// Reads an sfixed32 field value from the input.
+        /// </summary>   
+        public int ReadSFixed32()
+        {
+            return (int)ReadRawLittleEndian32();
+        }
+
+        /// <summary>
+        /// Reads an sfixed64 field value from the input.
+        /// </summary>   
+        public long ReadSFixed64()
+        {
+            return (long)ReadRawLittleEndian64();
+        }
+
+        /// <summary>
+        /// Reads an sint32 field value from the input.
+        /// </summary>   
+        public int ReadSInt32()
+        {
+            return CodedInputStream.DecodeZigZag32(ReadRawVarint32());
+        }
+
+        /// <summary>
+        /// Reads an sint64 field value from the input.
+        /// </summary>   
+        public long ReadSInt64()
+        {
+            return CodedInputStream.DecodeZigZag64(ReadRawVarint64());
+        }
+
+        /// <summary>
+        /// Reads a length for length-delimited data.
+        /// </summary>
+        /// <remarks>
+        /// This is internally just reading a varint, but this method exists
+        /// to make the calling code clearer.
+        /// </remarks>
+        public int ReadLength()
+        {
+            return (int)ReadRawVarint32();
+        }
+
+        /// <summary>
+        /// Peeks at the next tag in the input. If it matches <paramref name="tag"/>,
+        /// the tag is consumed and the method returns <c>true</c>; otherwise, the
+        /// input is left in the original position and the method returns <c>false</c>.
+        /// </summary>
+        public bool MaybeConsumeTag(uint tag)
+        {
+            uint previousTag = lastTag;
+            long consumed = reader.Consumed;
+
+            uint newTag = ReadTag();
+            if (newTag == tag)
+            {
+                // Match so consume tag
+                return true;
+            }
+
+            // No match so rewind
+            long rewindCount = reader.Consumed - consumed;
+            if (rewindCount > 0)
+            {
+                reader.Rewind(rewindCount);
+                lastTag = previousTag;
+            }
+
+            return false;
+        }
+
+        #endregion
+
```

Comment:
note that in the meantime, methods for accelerated reading Wrapper values have been added CodedInputStream - let's reimplement them or least add a TODO for adding them.

Reply:
I added them